File: ui/src/renderer/src/bindings/api/fs/fs.ts
import { invoke, invokeAsync, isTauri } from '@/bindings/core';

export type SelectDialogType = 'openFile' | 'openFiles' | 'openFolder' | 'openFolders' | 'saveFile';

export interface SelectDialogFilter {
  name: string;
  extensions: string[];
}

export interface SelectDialogOptions {
  type: SelectDialogType;
  title?: string | null;
  filters?: SelectDialogFilter[] | null;
  defaultDirectory?: string | null;
  defaultFileName?: string | null;
}

export const fs = {
  exists: async (path: string): Promise<boolean> => {
    return (await invoke('fs.exists', { path }))!;
  },
  isDir: async (path: string): Promise<boolean> => {
    return (await invoke('fs.is_dir', { path }))!;
  },
  isFile: async (path: string): Promise<boolean> => {
    return (await invoke('fs.is_file', { path }))!;
  },
  readToString: async (path: string): Promise<string> => {
    return (await invoke('fs.read_to_string', { path }))!;
  },
  readBinary: async (path: string): Promise<number[]> => {
    return (await invoke('fs.read_binary', { path }))!;
  },
  createDir: async (path: string): Promise<void> => {
    return (await invoke('fs.create_dir', { path }))!;
  },
  createDirAll: async (path: string): Promise<void> => {
    return (await invoke('fs.create_dir_all', { path }))!;
  },
  createFile: async (path: string, contents: string): Promise<void> => {
    return (await invoke('fs.create_file', { path, contents }))!;
  },
  delete: async (path: string, trash: boolean): Promise<void> => {
    return (await invoke('fs.delete', { path, trash }))!;
  },
  move: async (srcPath: string, targetPath: string, override: boolean): Promise<void> => {
    return (await invoke('fs.move', { srcPath, targetPath, override }))!;
  },
  copy: async (srcPath: string, targetPath: string, override: boolean): Promise<void> => {
    return (await invoke('fs.copy', { srcPath, targetPath, override }))!;
  },
  write: async (path: string, contents: string): Promise<void> => {
    return (await invoke('fs.write', { path, contents }))!;
  },
  writeBinary: async (path: string, buffer: ArrayBuffer): Promise<void> => {
    if (window.api) {
      window.api.utils.writeBinaryFile(path, buffer);
    } else if (isTauri) {
      throw new Error('todo tauri writeBinary');
    }
  },
  showInFolder: async (path: string): Promise<void> => {
    return (await invoke('fs.show_in_folder', { path }))!;
  },
  showSelectDialog: async (
    options: SelectDialogOptions,
  ): Promise<string | string[] | null | undefined> => {
    return (await invokeAsync('fs.show_select_dialog', options))!;
  },
  saveImageUrlToFile: async (imageUrl: string, filePath: string): Promise<void> => {
    return (await invokeAsync('fs.save_image_url_to_file', {
      imageUrl,
      filePath,
    }))!;
  },
};

File: ui/flutter/lib/src/bindings/api/fs/fs.dart
import 'package:lonanote/src/bindings/api/path/path.dart';
import 'package:lonanote/src/common/config/app_config.dart';
import 'package:lonanote/src/common/log.dart';

import '../../bindings.dart';

class RustFs {
  static bool exists(String path) {
    return Bindings.invoke(key: "fs.exists", value: {"path": path});
  }

  static bool isDir(String path) {
    return Bindings.invoke(key: "fs.is_dir", value: {"path": path});
  }

  static bool isFile(String path) {
    return Bindings.invoke(key: "fs.is_file", value: {"path": path});
  }

  static String readToString(String path) {
    return Bindings.invoke(key: "fs.read_to_string", value: {"path": path});
  }

  static List<int> readBinary(String path) {
    final res = Bindings.invoke(key: "fs.read_binary", value: {"path": path});
    final list = res as List<dynamic>;
    return list.cast<int>();
  }

  static void createDir(String path) {
    Bindings.invoke(key: "fs.create_dir", value: {"path": path});
  }

  static void createDirAll(String path) {
    Bindings.invoke(key: "fs.create_dir_all", value: {"path": path});
  }

  static void createFile(String path, String contents) {
    Bindings.invoke(
        key: "fs.create_file", value: {"path": path, "contents": contents});
  }

  static void delete(String path, bool trash) {
    Bindings.invoke(key: "fs.delete", value: {"path": path, "trash": trash});
  }

  static void move(String srcPath, String targetPath, bool override) {
    Bindings.invoke(key: "fs.move", value: {
      "srcPath": srcPath,
      "targetPath": targetPath,
      "override": override
    });
  }

  static void copy(String srcPath, String targetPath, bool override) {
    Bindings.invoke(key: "fs.copy", value: {
      "srcPath": srcPath,
      "targetPath": targetPath,
      "override": override
    });
  }

  static void write(String path, String contents) {
    Bindings.invoke(
        key: "fs.write", value: {"path": path, "contents": contents});
  }

  static Future<void> saveImageUrlToFile(
      String imageUrl, String filePath) async {
    await Bindings.invokeAsync(
        key: "fs.save_image_url_to_file",
        value: {"imageUrl": imageUrl, "filePath": filePath});
  }

  static void test() async {
    if (!AppConfig.isDebug) return;

    final testFilePath = "${RustPath.getCacheDir()}/test.txt";
    write(testFilePath, "test123");
    final fileText = readToString(testFilePath);
    final binary = readBinary(testFilePath);
    logger.i("RustFs.exists = ${exists(testFilePath)}");
    logger.i("RustFs.readToString = $fileText");
    logger.i("RustFs.readBinary = ${binary.length}");
    logger.i("RustFs.isFile = ${isFile(testFilePath)}");

    final testDirPath = "${RustPath.getCacheDir()}/testDir";
    createDirAll(testDirPath);
    logger.i("RustFs.createDirAll = ${exists(testDirPath)}");
    logger.i("RustFs.isDir = ${isDir(testDirPath)}");
    delete(testDirPath, false);

    final copyFilePath = "${RustPath.getCacheDir()}/copyText.txt";
    copy(testFilePath, copyFilePath, true);
    logger.i("RustFs.copy = ${exists(copyFilePath)}");

    final moveFilePath = "${RustPath.getCacheDir()}/moveText.txt";
    copy(testFilePath, moveFilePath, true);
    logger.i("RustFs.move = ${exists(moveFilePath)}");

    final imgPath = "${RustPath.getCacheDir()}/img.png";
    await saveImageUrlToFile(
        "https://raw.githubusercontent.com/lona-labs/lonanote/refs/heads/main/resources/icons/icon.png",
        imgPath);
    logger.i("RustFs.saveImageUrlToFile = ${exists(imgPath)}");

    delete(testFilePath, false);
    delete(copyFilePath, false);
    delete(moveFilePath, false);
    delete(imgPath, false);
    logger.i("RustFs.delete = success");
  }
}
